python/lsst/daf/butler/_butler.py

```diff
@@ -1179,62 +1167,41 @@ def put(
         TypeError
             Raised if the butler is read-only or if no run has been provided.
         """
-        if isinstance(datasetRefOrType, DatasetRef) and datasetRefOrType.id is not None:
+        if isinstance(datasetRefOrType, DatasetRef):
             # This is a direct put of predefined DatasetRef.
             log.debug("Butler put direct: %s", datasetRefOrType)
+            # _importDatasets ignores existing dataset ref and just returns an
+            # original ref.
```

Comment:
The comment is confusing, I'll update it. What I meant is that if the dataset with the same UUID, dataset type and dataId is already in the registry then `_importDatasets` simply returns it and does not raise (so I need to check datastore and raise exception myself below). If a dataset with the same dataset type and dataId but different UUID is already in the registry, then `_importDatasets` will raise an exception (same for matching UUID but different other items).

I'm not very comfortable changing how `_importDatasets` works, there may be places that rely on current behavior?

Reply:
If `_importDatasets` already raises, does that mean the test a couple of lines down was always irrelevant?

Reply:
It probably is now, it could have been relevant for some corner case before (e.g. directPut for integer dataset ID into UUID registry)? I think we can remove it now.